xcode/Subconscious/Shared/Services/AddressBookService.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  AddressBookService.swift
+//  Subconscious
+//
+//  Created by Ben Follington on 14/3/2023.
+//
+
+import Foundation
+import Combine
+// temp
+import SwiftUI
+
+class AddressBookService {
+    private(set) var noosphere: NoosphereService
+    private var addressBook: [AddressBookEntry]?
+    
+    init(noosphere: NoosphereService) {
+        self.noosphere = noosphere
+        self.addressBook = nil
+    }
+    
+    /// Get the full list of entries in the address book.
+    /// This is cached after the first use unless requested using refetch.
+    /// If an error occurs producing the entries the resulting list will be empty.
+    func listEntries(refetch: Bool = false) -> AnyPublisher<[AddressBookEntry], Never> {
+        if let entries = addressBook {
+            if !refetch {
+                return Just(entries).eraseToAnyPublisher()
+            }
+        }
+        
+        return self.listPetnamesAsync()
+            .flatMap { follows in
+                return Publishers.MergeMany(follows.map { f in
+                    return self.getPetnameAsync(petname: f)
+                        .map { did -> AddressBookEntry? in
+                            guard let did = did else { return nil }
+                            // TODO: hardcoded dog-pfp
+                            return AddressBookEntry(pfp: Image("dog-pfp"), petname: f, did: did)
+                        }
+                        .compactMap { $0 }
+                })
+                .collect()
+                .eraseToAnyPublisher()
+            }
+            .catch { err in
+                Just([])
+            }
+            .map { entries in
+                // Here be dragons (mutation) to cache the result
+                self.addressBook = entries
```

Comment:
Is using a class + simple mutation like this the "Swift" way to cache in-memory?

Reply:
Yup, this is the way.